feat(ui): unify builder settings fields handling - AB-210

VladKrasny opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Refactor**
  * Streamlined internal state management across multiple builder field components by adopting a unified field view model, resulting in more consistent and maintainable field value handling.
  * Updated input bindings and event handlers to use a new reactive approach, reducing manual state synchronization and direct content mutation.
  * Adjusted default value handling for checkbox, object, and text fields to support more flexible defaults.
  * Improved internal logic for various field types (text, color, padding, shadow, width, etc.) for a more reliable and predictable user experience.
  * Centralized alias and blueprint key management using the new field view model composable for simplified editing and selection.
  * Enhanced tools and notes components to utilize reactive field view models for data updates, improving responsiveness and consistency.
  * Refined event extraction logic in settings to safely handle missing component definitions.

* **Style**
  * Updated layout of the Writer Resource ID field to use a flexbox layout for improved alignment.

* **Chores**
  * Minor updates to internal type definitions to standardize parameters for field view models.
  * Simplified JSON parsing and default value fallbacks in evaluation utilities for more consistent behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

VladKrasny committed:
feat(ui): unify builder settings fields handling - AB-210

VladKrasny commented on src/ui/src/builder/settings/BuilderFieldsAlign.vue:
```diff
@@ -24,7 +24,7 @@
 				v-if="mode == 'css'"
 				ref="freehandInputEl"
-				:value="component.content[fieldKey]"
+				:value="fieldViewModel"
```
In Vue templates refs are automatically unwrapped, so there is no need to explicitly use .value here. The current binding works as intended.

VladKrasny committed:
fix(BuilderFieldsShadow): add null safety check for parsedValue - AB-210

madeindjs commented on src/ui/src/builder/settings/BuilderFieldsAlign.vue:
```diff
@@ -155,8 +150,13 @@ const props = defineProps({
 
-const { componentId, fieldKey, direction } = toRefs(props);
-const component = computed(() => wf.getComponentById(componentId.value));
+const { direction } = toRefs(props);
+
+const fieldViewModel = useComponentFieldViewModel({
+	componentId: toRef(props, "componentId"),
+	fieldKey: toRef(props, "fieldKey"),
```
```suggestion
const { direction, componentId, fieldKey } = toRefs(props);

const fieldViewModel = useComponentFieldViewModel({
	componentId,
	fieldKey,
```

madeindjs commented on src/ui/src/builder/settings/BuilderFieldsKeyValue.vue:
```diff
@@ -81,33 +81,30 @@ const props = defineProps({
+const fieldViewModel = useComponentFieldViewModel({
+	componentId: toRef(props, "componentId"),
+	fieldKey: toRef(props, "fieldKey"),
+	defaultValue: "",
```
You are not using the default value here. Some KeyValue fields has a default value set

https://github.com/writer/writer-framework/blob/aa186380e32ba3ccc32253d7a84428e0c5de4fe1/src/ui/src/components/core/input/CoreCheckboxInput.vue#L74-L80

https://github.com/writer/writer-framework/blob/aa186380e32ba3ccc32253d7a84428e0c5de4fe1/src/ui/src/components/core/input/CoreSelectInput.vue#L64-L70

This is also true for some other types .

You need to be sure your pass `:default-value` everywhere in `BuilderSettingsProperties.vue`.

Or you can also compute the default value in the hook:

```ts
type Params = {
	componentId: MaybeRef<string>;
	fieldKey: MaybeRef<string>;
};

export function useComponentFieldViewModel(
	params: Params,
	dependencies?: Dependencies,
) {
	const wf = dependencies?.wf ?? inject(injectionKeys.core);
	const ssbm = dependencies?.ssbm ?? inject(injectionKeys.builderManager);

	const component = computed(() => wf.getComponentById(params.componentId.value));

	const def = computed(() => wf.getComponentDefinition(component.value?.type));

	const defaultValue = computed(() => def.value?.[params.field.value]?.default);
```

VladKrasny replied:
As usual you are right, but I did not pass the default value intentionally to preserve the current behavior (with no defaults).

However, I agree this looks more like a bug in the existing logic, so the reasonable action would be to include default value propagation as part of this change

madeindjs reviewed (changes requested):
Nice refactoring. I think I spotted an issue with the default value on some fields.

VladKrasny committed:
refactor: access props.direction directly to avoid unnecessary ref creation, since props are already reactive

VladKrasny committed:
fix(BuilderFieldsText): correctly handle default value when expected to dispaly variable binding mode

Previously, the component used its default value from the component definition even in cases where variable binding was expected, which led to incorrect initialization.

VladKrasny committed:
feat: automatically set default values inside the composable or via passed params, so all fields using it will now receive defaults even if they were not previously provided

VladKrasny committed:
fix(ui): ensure supportedEvents is defined before object operations

madeindjs commented on src/ui/src/renderer/useEvaluator.ts:
```diff
@@ -195,16 +195,15 @@ export function useEvaluator(wf: Core, secretsManager?: SecretsManager) {
+				return JSON.parse(evaluated);
 			} catch {
-				return JSON.parse(defaultValue ?? null);
+				return JSON.parse(defaultValue || null);
```
well seen

VladKrasny committed:
fix(useComponentFieldViewModel): avoid throw in computed by validating component existence in watch
